### ISSUE TITLE
[FR] First-class metric metadata: goal direction (and unit) for logged metrics

### ISSUE THREAD

**Issue by atty57** [mlflow/mlflow#25377]
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

Let users declare, per metric key, whether higher or lower values are better (plus an optional unit / display precision), stored as experiment-scoped metric metadata:

```python
mlflow.set_metric_metadata("val_loss", greater_is_better=False, unit="nats")
mlflow.log_metric("val_loss", 0.31)
```

MLflow already has this concept, but only in the evaluation layer: `EvaluationMetric` and `make_metric()` / `make_genai_metric()` carry `greater_is_better`. It never reaches tracking — `mlflow.entities.Metric` has no such field, `search_runs` cannot use it, and there are no references to it anywhere in the UI codebase.

Consumers, in increasing order of scope:

1. `search_runs` / `search_logged_models`: `order_by=["metrics.val_loss BEST"]` instead of the caller hard-coding `ASC`/`DESC`.
2. UI: mark the best value in a metric column, orient chart axes, and colour deltas in the run comparison view correctly (a lower loss should not render as a red regression).
3. Model Registry: "promote the best version by metric X" stops needing a direction argument at every call site.

### Motivation

> #### What is the use case for this feature?

Direction is a property of the *metric*, but today it lives in the head of whoever writes each query, and every consumer re-derives it:

- Anyone sorting for a best run writes `ASC` or `DESC` from memory, and gets it silently wrong for the loss-shaped half of their metrics — a wrong answer, not an error.
- The run-comparison UI cannot say which run is better, so it shows raw deltas and leaves the interpretation to the reader.
- Sweep/AutoML tooling built on MLflow ships its own direction config (Optuna, Ray Tune, and every hand-rolled `max(runs, key=...)` loop) because MLflow has nowhere to put it.

> #### Why is this use case valuable to support for MLflow users in general?

The information is already being written down — just outside MLflow, in code comments and naming conventions like `*_loss` vs `*_score`. Moving it into the tracking layer makes every existing surface (search, charts, comparison, registry promotion) able to answer "which of these is better?" without per-call-site configuration.

It is also additive and backwards compatible: metrics without metadata behave exactly as they do today.

### Details

A few open questions I would like design feedback on before implementing:

- **Scope.** Experiment-level metadata keyed by metric name seems like the smallest thing that works. Per-run metadata would fragment (two runs disagreeing about a metric's direction).
- **Storage.** A `metric_metadata` table in the SQLAlchemy store, mirrored in the file store. This touches the Abstract Store, so it probably warrants the `needs design` treatment.
- **Auto-population.** Should `mlflow.evaluate()` populate the metadata from the `greater_is_better` its built-in metrics already declare? That would make the feature self-populating for the most common metrics with no user action.
- **REST API surface.** A dedicated `MetricMetadata` message plus set/get endpoints, or fold it into experiment tags under a reserved prefix to avoid new REST surface? The latter is considerably cheaper if the appetite for schema changes is low.
- **Extensibility.** The same per-key record is the natural home for unit and display precision, which would also cover #17150 (metric precision in the UI) rather than special-casing it.

### What machine learning domain(s) is this feature request about?

`domain/platform`

### What area(s) of MLflow is this feature request about?

`area/tracking`, `area/uiux`

**Comment by harupy** [comment]
Thanks for the writeup. Before the design questions, could you survey how other tools (W&B, Neptune, Comet, ClearML, TensorBoard HParams) handle this, and whether they attach direction to the metric or to the optimization job?

**Comment by atty57** [comment]
@harupy Surveyed. Short version: **four of the five attach direction to the optimization job, none of them durably attach it to the metric — and the one that tried removed it.**

| Tool | Direction lives on | Surface |
|---|---|---|
| W&B | the job (sweep) | `metric: {name, goal: minimize\|maximize, target}` in the sweep config |
| Comet | the job | `Optimizer` config `spec: {metric, objective: "minimize"\|"maximize"}` |
| ClearML | the job | `HyperParameterOptimizer(objective_metric_sign="min"\|"max"\|"min_global"\|"max_global")` |
| Optuna (for reference) | the job | `create_study(direction=/directions=)` |
| Neptune | neither — query time | `fetch_runs_table(sort_by=..., ascending=...)`, plus a per-attribute aggregation (`min()`/`max()`/`last()`/`average()`); nothing is stored on the metric |
| TensorBoard HParams | neither | `MetricInfo` carries `name`, `display_name`, `description`, `dataset_type` and *no* direction; ordering is `ColParams.order: ORDER_ASC/DESC` per request |

Three details worth more than the table:

**W&B built this and walked it back.** `define_metric()` used to take `goal="minimize"|"maximize"`, paired with `summary="best"`. Both are deprecated and the current public signature is `define_metric(name, step_metric, step_sync, hidden, summary, overwrite)` — no goal. The wire protocol still carries `MetricRecord.MetricGoal {GOAL_UNSET, GOAL_MINIMIZE, GOAL_MAXIMIZE}`, so the field survived while the user-facing API for setting it did not. What replaced it is `summary="min"` / `summary="max"` — direction expressed as *which aggregation is the summary*, which is strictly more informative and needs no new concept.

**TensorBoard is the sharpest negative result.** `MetricInfo` is exactly the "metric metadata message" this FR proposes, it already ships `display_name` and `dataset_type`, and direction was still left out.

**ClearML shows direction alone is under-specified.** `min` vs `min_global` distinguishes "the final reported value" from "the best value at any step". A bare `greater_is_better` can't express that, and any UI that highlights a best value has to answer it.

So I'd narrow the FR rather than defend it as written:

- Drop the optimization-objective framing entirely. Direction-as-objective belongs to the tuning job, and MLflow users already get it from Optuna/`hyperopt`/sweeps. Reproducing it in the tracking layer is the part the prior art argues against.
- Keep only what's genuinely metric-scoped and is the actual pain: **the UI and `search_runs` cannot tell which end of a metric column is good**, so delta colouring and "best run" are guesses. That's a presentation/selection problem, not an optimization one.
- Follow W&B's landing spot and make the declaration an aggregation, not a goal — something like `best="min"` / `best="max"` — which answers ClearML's final-vs-global question in the same field.
- No new REST message and no new store table in the opening version: experiment-level metadata under a reserved tag prefix, read by the UI and by an `order_by` keyword. That keeps it out of `needs design` territory and off the Abstract Store.

Does that shape seem worth pursuing, or does the survey result read to you as "the ecosystem settled this deliberately" and the issue should close? Happy with either — the prior art genuinely leans your way on the original scope.
